lib/matplotlib/dates.py

```diff
@@ -1409,13 +1411,19 @@ def tick_values(self, vmin, vmax):
         ymin = self.base.le(vmin.year) * self.base.step
         ymax = self.base.ge(vmax.year) * self.base.step
 
-        ticks = [vmin.replace(year=ymin, **self.replaced)]
+        vmin = vmin.replace(year=ymin, **self.replaced)
+        try:
+            ticks = [vmin.astimezone(self.tz)]
+        except ValueError as e:
+            raise ValueError('naive datetime objects cannot be used '
+                             'with matplotlib for python < 3.6; ') from e
         while True:
             dt = ticks[-1]
             if dt.year >= ymax:
                 return date2num(ticks)
             year = dt.year + self.base.step
-            ticks.append(dt.replace(year=year, **self.replaced))
+            dt = dt.replace(year=year, **self.replaced).astimezone(self.tz)
```

Comment:
I think you actually don't want `astimezone` in this situation. You want `dt.replace(tzinfo=self.tz)`.

I could be wrong, but I think what you probably want to do is leave `tzinfo=tz` in the `self.replaced` dictionary, and in this place add:

```python
dt = dt.replace(year=year, **self.replaced)
if hasattr(self.tz, 'normalize'):
    dt = self.tz.normalize(dt, dst=True)
```

My intuition about the `tz` parameter of `YearLocator` is that it *should* be returning ticks at yearly intervals in the local time zone.

Reply:
Unfortunately, `normalize` doesn't work.  First `pytz.normalize` doesn't seem to have a dst kwarg.  But that asside, the following example is not at a time of year where DST should be an issue:

```import datetime
import pytz
import matplotlib.dates as mdates

tz = datetime.timezone(datetime.timedelta(hours=-5))

dt = datetime.datetime(2007, 1, 1, 0, 0, 0, tzinfo=tz)
yday = mdates.date2num(dt)
print(dt, yday, mdates.num2date(yday, tz=tz))

dt = datetime.datetime(2007, 1, 1, 0, 0, 0)
dt = dt.replace(tzinfo=tz)
yday = mdates.date2num(dt)
print(dt, yday, mdates.num2date(yday, tz=tz))

tz = pytz.timezone('America/New_York')

dt = datetime.datetime(2007, 1, 1, 0, 0, 0)
dt = dt.replace(tzinfo=tz)
yday = mdates.date2num(dt)
print(dt, yday, mdates.num2date(yday, tz=tz))

dt = datetime.datetime(2007, 1, 1, 0, 0, 0)
dt = dt.replace(tzinfo=tz)
dt = tz.normalize(dt)
yday = mdates.date2num(dt)
print(dt, yday, mdates.num2date(yday, tz=tz))

dt = datetime.datetime(2007, 1, 1, 0, 0, 0)
dt = dt.astimezone(tz=tz)
yday = mdates.date2num(dt)
print(dt, yday, mdates.num2date(yday, tz=tz))

dt = tz.localize(datetime.datetime(2007, 1, 1, 0, 0, 0))
yday = mdates.date2num(dt)
print(dt, yday, mdates.num2date(yday, tz=tz))
```

yields 
```
2007-01-01 00:00:00-05:00 732677.2083333334 2007-01-01 00:00:00-05:00
2007-01-01 00:00:00-05:00 732677.2083333334 2007-01-01 00:00:00-05:00
2007-01-01 00:00:00-04:56 732677.2055555555 2006-12-31 23:56:00-05:00
2006-12-31 23:56:00-05:00 732677.2055555555 2006-12-31 23:56:00-05:00
2007-01-01 03:00:00-05:00 732677.3333333334 2007-01-01 03:00:00-05:00
2007-01-01 00:00:00-05:00 732677.2083333334 2007-01-01 00:00:00-05:00
```

So, unless I'm doing something wrong (completely possibe) the only way to round-trip `pytz` is to use `localize` (which I'm *not* doing in this PR yet...)

Reply:
OK, new version replaces timezone info, unless it has a localize method, in which case it does that....  Basically the same code as used in `RRuleLocator`...

Reply:
Ah, yeah, I forgot that the way `normalize` is implemented it assumes that you started with a valid time and then did some arithmetic or something on it. Very dangerous.

The equivalent code with `localize` is this:

```python
dt = dt.replace(year=year, **self.replaced)
if hasattr(self.tz, 'localize'):
    dt = self.tz.localize(dt.replace(tzinfo=None), is_dst=True)
```

I think there are other ways to do it as well. Another option would be to define a module-level `_needs_pytz_localize` like this:

```python
def _needs_pytz_localize(tzinfo):
    global _needs_pytz_localize
    try:
        import pytz
        def _needs_pytz_localize(tzinfo):
            return isinstance(tzinfo, pytz.tzinfo.DstTzInfo)
    except ImportError:
        def _needs_pytz_localize(tzinfo):
            return False

    return _needs_pytz_localize(tzinfo)
```

Then instead of duck-typing you can specifically check for `pytz` time zones that need to be localized (fixed offsets don't).

The downside of this is that it imports pytz even if you're not using it. The upside is that other time-zone providers that use a proper Pythonic interface but offer compatibility with `pytz` will get end up opting in to the non-pythonic interface.

Reply:
Note that I don't know of any other time zone providers that do this, and `dateutil` never will, so I think that sort of duck typing is probably fine, just throwing it out there.

Reply:
I put a comment in the code that says we are doing this for `pytz`.  If someone trips across this with another timezone provider that has a `localize` thats not compatible, we can revisit?